lib/src/main/java/uk/gov/di/ipv/core/library/validation/TokenRequestValidator.java

```diff
@@ -1,37 +1,118 @@
+import uk.gov.di.ipv.core.library.exceptions.ClientAuthenticationException;
+import uk.gov.di.ipv.core.library.service.ConfigurationService;
 
-import java.text.ParseException;
+import java.time.OffsetDateTime;
+import java.util.Date;
 import java.util.Map;
+import java.util.Set;
 
 public class TokenRequestValidator {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(TokenRequestValidator.class);
+    private static final String CLIENT_ASSERTION_PARAM = "client_assertion";
+    private static final String CLIENT_ID_PARAM = "client_id";
+    private static final String NONE = "none";
+    private static final String JWT = "jwt";
 
-    public ValidationResult<ErrorObject> validateExtractedJwt(String requestBody) {
+    private final ConfigurationService configurationService;
 
-        Map<String, String> stringMap = RequestHelper.parseRequestBody(requestBody);
+    private final ClientAuthenticationVerifier<Object> verifier;
 
-        try {
-            SignedJWT clientJwt =
-                    SignedJWT.parse(String.valueOf(stringMap.get("client_assertion")));
+    private String clientId;
 
-            JWTClaimsSet claimsSet = clientJwt.getJWTClaimsSet();
+    public TokenRequestValidator(ConfigurationService configurationService) {
+        this.configurationService = configurationService;
+        this.verifier = getClientAuthVerifier(configurationService);
+    }
+
+    public void authenticateClient(String requestBody, Map<String, String> queryParams)
+            throws ClientAuthenticationException {
+        if (!queryParams.containsKey(CLIENT_ASSERTION_PARAM)) {
+            if (!queryParams.containsKey(CLIENT_ID_PARAM)) {
+                LOGGER.error(
+                        "Missing either client_assertion or client_id values in request. Failed to establish client_id value.");
+                throw new ClientAuthenticationException(
+                        "Unknown client, no client_id value or client_assertion jwt found in request");
+            } else {
+                clientId = queryParams.get(CLIENT_ID_PARAM);
+
+                String clientAuthenticationMethod = getClientAuthenticationMethod(clientId);
 
-            if (claimsSet != null) {
-                return ValidationResult.createValidResult();
+                if (clientAuthenticationMethod.equals(JWT)) {
+                    LOGGER.error("Missing client_assertion jwt for configured client {}", clientId);
+                    throw new ClientAuthenticationException(
+                            String.format(
+                                    "Missing client_assertion jwt for configured client '%s'",
+                                    clientId));
+                }
             }
-        } catch (ParseException e) {
-            LOGGER.error("Unable to parse Claims set {} ", e.getMessage());
-            return new ValidationResult<>(false, OAuth2Error.INVALID_CLIENT);
+        } else {
```

Comment:
This is quite hard to follow through all the branches - could it be refactored to a well name private method?

Reply:
Refactor into 2 distincts smaller method call